python/cudf/cudf/_lib/copying.pyx

```diff
@@ -735,3 +737,196 @@ def segmented_gather(Column source_column, Column gather_map):
+            "strides": None,
+            "typestr": "|u1",
+            "version": 0
+        }
+        return intf
+
+    @property
+    def metadata_ptr(self):
+        return int(<uintptr_t>self.c_metadata_ptr())
+
+    @property
+    def metadata_size(self):
+        return int(self.c_metadata_size())
+
+    @property
+    def gpu_data_ptr(self):
+        return int(<uintptr_t>self.c_gpu_data_ptr())
+
+    @property
+    def gpu_data_size(self):
+        return int(self.c_gpu_data_size())
+
+    def serialize(self):
+        header = {}
+
+        header["column-names"] = self.column_names
+        header["index-names"] = self.index_names
+        header["gpu-data-ptr"] = self.gpu_data_ptr
+        header["gpu-data-size"] = self.gpu_data_size
+        header["metadata"] = list(
+            <uint8_t[:self.metadata_size]>self.c_metadata_ptr()
+        )
+
+        return header, []
+
+    @classmethod
+    def deserialize(cls, header, frames):
+        cdef PackedColumns p = PackedColumns.__new__(PackedColumns)
+
+        dbuf = DeviceBuffer(
+            ptr=header["gpu-data-ptr"],
+            size=header["gpu-data-size"]
+        )
+
+        cdef cpp_copying.packed_columns data_
+        data_.metadata_ = move(
+            make_unique[cpp_copying.metadata](
+                move(<vector[uint8_t]>header["metadata"])
+            )
+        )
+        data_.gpu_data = move(dbuf.c_obj)
+
+        p.c_obj = move(data_)
+        p.column_names = header["column-names"]
+        p.index_names = header["index-names"]
+
+        return p
+
+    @staticmethod
+    cdef PackedColumns c_from_py_table(Table input_table, keep_index=True):
+        """
+        Construct a PackedColumns object from a ``cudf.DataFrame``.
+        """
+        from cudf.core import RangeIndex
+
+        cdef PackedColumns p = PackedColumns.__new__(PackedColumns)
+
+        if keep_index and not input_table.index.equals(
+            RangeIndex(start=0, stop=len(input_table), step=1)
+        ):
+            input_table_view = input_table.view()
+            p.index_names = input_table._index_names
+        else:
+            input_table_view = input_table.data_view()
+
+        p.c_obj = move(cpp_copying.pack(input_table_view))
+        p.column_names = input_table._column_names
+
+        return p
+
+    cdef Table c_unpack(self):
+        return Table.from_table_view(
+            cpp_copying.unpack(self.c_obj),
+            self,
+            self.column_names,
+            self.index_names
+        )
+
+    cdef const void* c_metadata_ptr(self) except *:
+        return self.c_obj.metadata_.get()[0].data()
+
+    cdef size_t c_metadata_size(self) except *:
+        return self.c_obj.metadata_.get()[0].size()
+
+    cdef void* c_gpu_data_ptr(self) except *:
+        return self.c_obj.gpu_data.get()[0].data()
+
+    cdef size_t c_gpu_data_size(self) except *:
+        return self.c_obj.gpu_data.get()[0].size()
+
+
+def pack(Table input_table, keep_index=True):
```

Comment:
Are these meant to be public APIs? If not, we should mention "internal" at doc, or make docs less like public API docs.

Reply:
In general the goal here was to use pack/unpack optionally as part of DataFrame serialization internally, but I'm not 100% if there's functionality in using it outside of cuDF. I'll check in on this and update the docs accordingly.

Reply:
I'm going to end up going with minimal docs here since I don't think these are mean to be public APIs - if that change in the future, we can always add them back later